### ISSUE TITLE
Chatroom No. 2

### ISSUE THREAD

**Issue by NikoKrause** [NikoKrause/CinnVIIStarkMenu#9]
Chatroom No. 1 https://github.com/NikoKrause/CinnXPStarkMenu/issues/6 has too many comments. I need centuries to scroll to the bottom. So I close No. 1. New room for new chats ...
@cinnamoner
@feren
@Elbullazul
@BigDaddyonTech

**Comment by NikoKrause** [comment]
@BigDaddyonTech
Check the Readme file ;)

**Comment by cinnamoner** [comment]
@NikoKrause Inasmuch as the stock Cinnamon menu fitting the needs of an average user you might be correct. The thing is I'm not an average user or average anything. I never was an average Joe, nor will I ever be. I'm notoriously hard to please. I wouldn't even consider using Cinnamon without StakMenu which emulates the mechanics/physics of the Vista/7 menu. To me, that particular design is the best ever envisioned. I'm tremendously grateful for the effort you've invested in Stark and seemingly I'm not the only one raving about it.

**Comment by BigDaddyLinux** [comment]
> Check the Readme file ;)

@NikoKrause That's awesome!! You da man

**Comment by BigDaddyLinux** [comment]
I am not a fan of the stock menu for multiple reasons.
1. No customization
2. Favorites has no text or option to use text....hence the reason I like stark
3. No customization
4. No customization
:)

**Comment by NikoKrause** [comment]
There is a new option in the contextMenu of the original menu. Something with bumblebee/nvidia.
I don't have nvidia/bumblebee. Is somebody of you using this or can check it?

**Comment by cinnamoner** [comment]
@BigDaddyonTech You're darn right. Those are the reasons I don't like the stock menu either. At any rate you've nailed it way better than I had.

**Comment by NikoKrause** [comment]
@guys
I did some code cleanup. So I basically compared the CinnXPStarkMenu code with the original menu code, which means I copied everything what was the same, but didn't looked the same by format.

This will make it easier to maintain the applet in the future, i.e. repair the applet if it's broken and adapt bug fixes and new features from the original menu (upstream).

It was a lot of code. Over 1000 lines of code. And i hope I didn't broke anything. 
I know for sure, that I catched a minor fix (the enable pathsystem is now navigable, it didn't worked before).

I would appreciate, if you test the hell out of it and tell me if you find any regressions. It means if you find something, what is not working anymore, but worked before.

Also it would be nice to know, if it works for LinuxMint 17.x?

Thanks in advance for testing and feedback ;)

P.S.: I removed the Help button from the sidebar. I didn't used it and the help section, which was opened by this button, doesn't seem really helpful to me, e.g. it's only in English.

**Comment by BigDaddyLinux** [comment]
@NikoKrause 
I'll be glad to test it out but it may be a few days

**Comment by cinnamoner** [comment]
@NikoKrause I've been following the action. I think it was a good idea to put it against stock. In theory the code should be cleaner now. I'll try it out on 17.2 and 18.0. I'm avoiding 17.3 because of the flicker bug present in that release (guess they hadn't payed attention 'cause their mind was already on 18.0 which, just between you and me, is not a successful build either; perhaps Serena will be better). Thanks Niko.

**Comment by NikoKrause** [comment]
There's no hurry for testing.
I fixed some broken stuff and added 1 or 2 minor new features.
The biggest change was the keynavigation improvement. (The quicklinks sidebar still isn't navigable. I'm not working on that right now.)
But I adapted the improvement of the keynav from upstream, it means the keys to navigate are:
- As always LEFT, RIGHT, UP, DOWN
- PAGE_UP, PAGE_DOWN, TAB, SHIFT+TAB
- MENU_KEY or CTRL+ENTER to open contextMenu on application/favorite/recent buttons
- CTRL+UP or CTRL+DOWN in the favorites box to change the order of the favorites. 
- DELETE-KEY in the favorites box to remove a favorite from the favorites box.

@cinnamoner 
I'd be happy to hear from you if it works on 17.2/18.0.
Could you please tell the Cinnamon version as well after testing.
I tested it on Cinnamon 2.8.8 and Cinnamon 3.2.1, so I think it should also work in between.
But I would like to know if it works on versions < 2.8.8
The reason is, I want to move it to multiversion support, because API has changed a lot for 3.2. For example vertical panels support will break the applet for minor versions.

**Comment by cinnamoner** [comment]
@NikoKrause Okay Niko. I intend to test in this weekend. Thanks.

**Comment by NikoKrause** [comment]
@BigDaddyonTech
I really like your video and your voice. You have a calm and pleasant voice. 
I'd like you to ask, if you are interested in making another video about CinnXPStarkMenu, when I'm finished with improving it for Cinnamon 3.2?

Showing all options in it. Also explaining the keynavigation possibilities (some of them are not that obvious).

It's also not very clear, how to add new buttons to the Quicklinks Sidebar (you mention that in your current video). I think it's pretty easy if you see it once, but it's not very obious, if you don't know how to do it. So explaining it in a video, would help a lot of people, who want to use this Option. (I will send you instructions on how to do this as simple as possible, if you are intersted in making another video.)

Also you mentioned the absence of this applet on spices.linuxmint.com. I'll upload it to the page, when I'm finished and before you are starting to make the video.

Waiting impatiently for your response ;)

**Comment by dominichayesferen** [comment]
@NikoKrause Is 3.2 fully in Cinnamon Nightly yet?

**Comment by NikoKrause** [comment]
@feren 
What do you mean by fully? Systeminformation shows 3.2.1. The guy also added the xapps, because of some unsolved dependencies. So you can check them out as well.

**Comment by dominichayesferen** [comment]
@NikoKrause Beforehand, it wouldn't fully install due to dependency issues (Cinnamon would be removed by Update Manager)

**Comment by NikoKrause** [comment]
Yes that issue was solved.

**Comment by cinnamoner** [comment]
@NikoKrause Testing is coming, don't worry. Just a little short on time. By the way, if you're still interested in Configurable, the Makulu fork seems to work okay on 2.8.8. Due to lack of founding the project seems to have halted. But you can download the latest beta from SourceForge.

**Comment by NikoKrause** [comment]
No pressure on Testing. I still have some stuff to push.

Do I have to download the whole Makulu OS?

**Comment by cinnamoner** [comment]
> Do I have to download the whole Makulu OS?

Truth to tell, I haven't downloaded. I've just watched the video. I'll provide (for convenience) the link in a minute. It will lead to both the video and the OS itself. It's up to you whether you want to download the iso.

**Comment by cinnamoner** [comment]
> No pressure on Testing. I still have some stuff to push.

Okay, no pressure. Take all the time you need.

**Comment by NikoKrause** [comment]
Sorry, I think that came out wrong. I don't want to pressure you, to test the new version. Not the other way  round. :)

**Comment by cinnamoner** [comment]
Goes both ways :) I need some time. You need some time. Everything's a okay.

**Comment by BigDaddyLinux** [comment]
> @BigDaddyonTech
> I really like your video and your voice. You have a calm and pleasant voice. 
> I'd like you to ask, if you are interested in making another video about CinnXPStarkMenu, when I'm finished with improving it for Cinnamon 3.2?
> 
> Showing all options in it. Also explaining the keynavigation possibilities (some of them are not that obvious).
> 
> It's also not very clear, how to add new buttons to the Quicklinks Sidebar (you mention that in your current video). I think it's pretty easy if you see it once, but it's not very obious, if you don't know how to do it. So explaining it in a video, would help a lot of people, who want to use this Option. (I will send you instructions on how to do this as simple as possible, if you are intersted in making another video.)
> 
> Also you mentioned the absence of this applet on spices.linuxmint.com. I'll upload it to the page, when I'm finished and before you are starting to make the video.
> 
> Waiting impatiently for your response ;)

@NikoKrause I would be happy to do another video on your menu. All you have to do is give me time to test it and give me the important points you want me to cover. I think the menu is awesome so any features that I don't know about, I would love to know.
I have not had time to test it as of yet but I will.

**Comment by NikoKrause** [comment]
I'm glad to hear you're interested @BigDaddyonTech 
I'll come back to you, when I'm feeling ready.

**Comment by NikoKrause** [comment]
@mintystark
I forked your StarkMenu. I can't find any mentioning about license. 
Could you please tell me, which License I can use? I thought of GPL.
(I think the base for the StarkMenu was the original menu-applet of Cinnamon, which is licensed und er GPL. Am I right?)
Regards
NikoKrause

**Comment by mintystark** [comment]
I actually forked it from someone else.  
It was forked from a "Windows 7+Cinnamon - Menu" Applet.
The creators name was "physics"

I think he just modified the main Cinnamon Menu at the time.  Quite a few years ago.

**Comment by cinnamoner** [comment]
@physics Do you have a Git account?

**Comment by dominichayesferen** [comment]
While it won't affect CinnXPStarkMenu, I thought it was worthy of mentioning this:
https://github.com/linuxmint/Cinnamon/issues/5921

**Comment by dominichayesferen** [comment]
@NikoKrause Tried adding the latest version of CinnXPStarkMenu from Git to a Vertical Panel, won't work

**Comment by dominichayesferen** [comment]
...it still says not suitable for Vertical Panels... :/

**Comment by NikoKrause** [comment]
@feren 
I know. They changed the code for that. It's not a big deal to fix, but it will break the menu for Cinnamon < 3.2
I'll fix it, when I switch to multiversion support. Hope you are patient enough.

@cinnamoner 
Yes, he has a Github account. Only on another name
https://cinnamon-spices.linuxmint.com/applets/view/84

**Comment by NikoKrause** [comment]
@jake-phy
Hello,
I forked the StarkMenu by mintystark, which is a fork of your Win7 menu emulator.
Apparently no one attached a license to it. I think your menu is a fork of the original cinnamon menu-applet, because of the similarities in the code base.
Can I assume it's the same license GPL?
Regards
NikoKrause

**Comment by NikoKrause** [comment]
@feren 
If you can't wait. It's only 4 lines of code. 3 deletions and 1 addition.
That's the commit:
https://github.com/linuxmint/Cinnamon/commit/159f1f1a62f430caf3c1eb72081c319d6c00d5e5

**Comment by cinnamoner** [comment]
@NikoKrause I suggest to go slowly. Multiversion is great but you should wait until Serena comes out. Just an opinion though.
***
As of Configurable, I'm willing to download the Lindoz build. Maybe reverse engineering could work. I have no idea. What's your opinion?

**Comment by ghost** [comment]
@cinnamoner, anyone can fix the menu if he want. I say it's not difficult fix it from the spice version, in any case, I'm pretty sure, the menu have memory leaks. If you find my first commit before 2.8.8, the original problem it's fixed there, but i do more things in that commit and this is the cause of more problems. Anyway definitely the menu will not work on 3.2 where sure there are a lot of changes.

**Comment by dominichayesferen** [comment]
@lestcape Well, for now, at least we can enjoy it for a bit longer, who knows, someone could somehow port it to 3.2 if they tried...

Also, hey! I guess Cinnamoner called you over here then, also, GJ at the LIM Applet too, works really well over here

**Comment by cinnamoner** [comment]
@lestcape Thanks for the answer. Jacque (Makulu) has just told us (to feren an me) he has it working on 3.0.7. Also just said he may stick with 3.0.7. Which confirms what you're telling me, namely 3.2 is a big bump in the road for Configurable. Thanks again.

**Comment by cinnamoner** [comment]
Obviously, there's a difference between *working* and *all leaks solved*. But I'm out of my league here. I'm mainly a guy who establishes contact between developers whenever needed (or not :))

**Comment by NikoKrause** [comment]
@feren
Do you remember the quick&dirty hack for the quit/logout/lockscreen buttons I did for the Vista theme?
It's a huge mess. There is a lot of the same code 4 times. I want to tidy up.

There was a problem with themeing the quit button in horizontal and vertical, if I remember it right, because they used the same style-class.

Do you want for each button an extra style class?
starkmenu-quit-horizontal-button
starkmenu-quit-vertical-button
starkmenu-quit-dropdown-button

And resp. for logout and lockscreen buttons?

**Comment by Elbullazul** [comment]
@nikokrause is rather have different classes for different layouts.  It would prevent having weird looking buttons in vertical mode

**Comment by dominichayesferen** [comment]
@NikoKrause Agreed with @Elbullazul here, also, your menu works fine otherwise in 3.2...

**Comment by cinnamoner** [comment]
@NikoKrause I still think StarkMenu is sufficient to focus on. The Configurable menu talk was more like a parenthesis than anything else.
***
Obviously, I like you're focusing on making the code clean.

**Comment by NikoKrause** [comment]
@Elbullazul @feren 
I changed the style-class-names:
```
starkmenu-arrow-dropdown-button

starkmenu-quit-dropdown-button
starkmenu-quit-horizontal-button
starkmenu-quit-vertical-button

starkmenu-logout-horizontal-button
starkmenu-logout-vertical-button

starkmenu-lockscreen-horizontal-button
starkmenu-lockscreen-vertical-button
```

Please let me know, when you updated your themes.

**Comment by cinnamoner** [comment]
@Elbullazul @feren ^ ?

**Comment by dominichayesferen** [comment]
Not yet here, been busy configuring my Network Printer (HP OfficeJet) for feren OS...

**Comment by NikoKrause** [comment]
@feren 
Will the configurable menu will be included in you new ferenOS release?

**Comment by dominichayesferen** [comment]
Yeah, but locally, because that's the only way it'll work, and then I'll see if it'll work in 3.2, also, don't worry, CinnXPStarkMenu will remain the menu of choice in Themer... ;)

**Comment by NikoKrause** [comment]
@feren 
You want to fork Slingshot? Which one? There are two on spices.
I once tried one of them. It was incredibly slow!

**Comment by cinnamoner** [comment]
@feren Happy to read you're keeping Stark. Best ever applet for Cinnamon. Even better than Sling. That's my opinion right there.

**Comment by dominichayesferen** [comment]
@NikoKrause The only one listed in Cinnamon Spices on Cinnamon, the one by jfarthing

**Comment by NikoKrause** [comment]
There are two:
https://cinnamon-spices.linuxmint.com/applets/view/122
https://cinnamon-spices.linuxmint.com/applets/view/197

**Comment by dominichayesferen** [comment]
The port of Slingshot

**Comment by NikoKrause** [comment]
The one I tried was that from jfarthing.
I think it was really slow. But if you are interested in forking it, I once created a pull request, which sorted the categories. He never seemed interested in it. So I closed it. You can still find it on his repo.

**Comment by dominichayesferen** [comment]
@NikoKrause Is it just me, or is Minty Colours REALLY broken on 3.0?

**Comment by NikoKrause** [comment]
@feren 
What do you mean by Minty Colours?

@guys
I just got a little bit excited. The menu is gotten a little bit better now. And for the first time, I don't have the feeling to drop the dropdown quit menu :)
It's pretty nice now :)

**Comment by dominichayesferen** [comment]
@NikoKrause Minty Colours = The Theme on Cinnamon Spices

Noice

**Comment by dominichayesferen** [comment]
@NikoKrause It looks like you're running Cinnamon while there's NO cinnamon themes available at all when you try to use Minty Colours in 3.0...

**Comment by dominichayesferen** [comment]
UPDATE: Never mind, it loads cinnamon when I tried it this time...

**Comment by cinnamoner** [comment]
@NikoKrause Nope! You don't have to download the entire Makulu iso. Feren did it for you. Seems to work on Cinnamon 3.0.0 - 3.0.7. Have fun :)
[configurableMenu.lestcape.tar.gz](https://github.com/NikoKrause/CinnXPStarkMenu/files/611216/configurableMenu.lestcape.tar.gz)

**Comment by NikoKrause** [comment]
I had the following bug on my bucket list:
> After removing favorite with largest name, the favorites box should resize while menu is still open, but it doesn't.

I just noticed, that the bug is gone. And I have no idea, when or why this happened o.O

**Comment by dominichayesferen** [comment]
@NikoKrause Another bug, your menu doesn't have a scrolling area when the user has so many favourites, it won't fit in the menu

**Comment by cinnamoner** [comment]
@NikoKrause To which menu have you reffered to? I'm confused. ⬆

**Comment by dominichayesferen** [comment]
@cinnamoner I think his menu...

**Comment by cinnamoner** [comment]
@feren *this* doesn't mean anything. Whenever reference is made to any menu, one should give its name. I kinda blame myself for having started talking about Configurable in a StarkMenu thread. Yes, it takes loner to post in clear way, but not posting in clear way takes even longer because it breaks the flow of communication.

**Comment by NikoKrause** [comment]
@feren 
> @NikoKrause Another bug, your menu doesn't have a scrolling area when the user has so many favourites, it won't fit in the menu

I'm aware of it. I do not use a lot of Favorites. So never saw it as a bug ;) 
It could be solved by resizing the icons.

I think the stock menu is resizing the favIcons. StarkMenu is using custom size. Have to look into in, someday.

**Comment by dominichayesferen** [comment]
@NikoKrause Maybe you could see if its possible to exploit the scrolling pane of All Applications to make another scrolling pane carrying favourites?

**Comment by cinnamoner** [comment]
@NikoKrause Have you had the time to play with the Makulu Configurable menu? If so, have you noticed any noteworthy annoyances in it?

**Comment by ghost** [comment]
@NikoKrause i think that a visible scroll in favorites don't  look good in some context, if you want to hide the scroll handle:
```
this.scroll.get_vscroll_bar().visible = false;
this.scroll.get_hscroll_bar().visible = false;
```

**Comment by NikoKrause** [comment]
@lestcape 
Thanks for the hint. 

I'm not sure if I would like the favoritesBox to be scrollable. When I remember it right, the original menu resizes the favorites icons, if you are getting more and more. I never pushed it to the limit though. Is the number of favorites limited?

@feren What's the default behaviour in Windows 7 in this case?

@lestcape again: Maybe you can help me out with Drag&Drop of the favorites buttons. The original menu prevents placing an favorite before or after itself.      https://github.com/linuxmint/Cinnamon/blob/master/files/usr/share/cinnamon/applets/menu%40cinnamon.org/applet.js#L1040
 It doesn't work for my applet. I assume it's because my favoritesButtons have contextMenu.
I couldn't figure out, how to fix it for my menu. Do have any idea?

@cinnamoner 
Thanks for the file. Hadn't time yet to check it out. 
Regarding the Win10 menu on Plasma. Watched the video. It looks great.
The nearest you can get with StarkMenu is using Mate-Menu layout and Icons only on the Quicklinks Sidebar. It's more a job for configurable.

**Comment by ghost** [comment]
@NikoKrause 

You are right, the favorites icons will be resize... I create configurable menu just for one reason: The possibility of display the cinnamon menu in a notebook (my notebook), so sure the height of the menu it's important to me. I don't know, but in configurable menu you can set a grid of icons... When i test it, the numbers of favorites was really big, and nothings happens.

Yes when you introduced more thing you will need to take this on account. Really 2*(number of menu items) as actually you need to take on account the menu. there are another way and this is not add the menu to the box, add the menu and only the menu that you need and when you need the menu, after close it, remove the menu from the box. In that way you will gain speed on the menu as clutter will not need to recompute the layout for a lot of invisible actors, and also you can estimate better the position of the actors in the box. But ofcourse this need probably a more big change in your code.

**Comment by Elbullazul** [comment]
@NikoKrause Windows stretches the menu by the top.  It makes it higher so it can display all favorites.

**Comment by NikoKrause** [comment]
@Elbullazul 
What happens, when the top is reached?

**Comment by dominichayesferen** [comment]
@NikoKrause Windows 10: Tiles becomes a scrollable area
Windows 7 and older: It simply doesn't show the rest of the Pinned Apps from what I've seen

**Comment by Elbullazul** [comment]
@nikokrause good question.  I'll investigate today and come back with the answer

**Comment by dominichayesferen** [comment]
@NikoKrause Is this normal behaviour for Cinnamon to sometimes not have an About option on Applets?
![screenshot from 2016-11-25 12-17-58](https://cloud.githubusercontent.com/assets/11057934/20625359/be76a5fc-b30a-11e6-9e67-322b09ca152e.png)
(BTW, that's when Slingshot doesn't scale its icon, otherwise, it was About, Configure and Remove)

**Comment by Elbullazul** [comment]
@feren probably the javascript applet just stopped execution when the applet got misconfigured

**Comment by Elbullazul** [comment]
@NikoKrause just like feren says, the older favorites don't get displayed, and only shows the top latest that fit in the menu.

**Comment by cinnamoner** [comment]
@NikoKrause Well, seems you've now got a crowd on your head. Things are really easy. StarkMenu is ment to emulate the Vista/7 menu. I don't think it's ment to emulate the Windows 10 menu (which is arguably not even a menu). I don't say experiments are not good but ... Yeah, I understand you don't wanna change the initial purpose of Stark (which is good). You've also provided a variant that tries to come close to the W10 menu. The whole thing started from that KDE tile menu, I know :)

**Comment by NikoKrause** [comment]
> StarkMenu is ment to emulate the Vista/7 menu

You are right. I thought about dropping the MATE-Menu layout.
Is anybody of you using this or has some objections to not remove this option?

It will make the keynavigation more complicated, when the sibebar has to be made navigable.

**Comment by ghost** [comment]
@NikoKrause take care when you have a lot of code. Always someone think that it's good to have a new ability in cinnamon, unify some behavior in one place, make an standardization, remove unused features, or just the most commons, a better look.

**Comment by cinnamoner** [comment]
@NikoKrause Keep it simple. Vista/7 only. I'd say to drop the Mate part. Thanks.

**Comment by cinnamoner** [comment]
@NikoKrause I agree with lestcape.

**Comment by ghost** [comment]
@feren i really don't know why you want to use configurable menu with memory leaks, but well:
https://github.com/lestcape/Configurable-Menu/commit/ea224597ae613aa86ba473df789d3c9a604e4e50

**Comment by NikoKrause** [comment]
Yeah, I implemented the MATE-layout back then, because the Shutdown-Menu behaved weird in some situations. Now that I fixed the shutdown menu and I am looking forward to keynavigation it seems better to drop MATE-layout. Just wanted to know if anybody is using it.

I also tested the favorites of the stock menu to the limits. At least I tried. My screen got to small.
It's ridiculous:
![favorites overkill](https://cloud.githubusercontent.com/assets/8415124/20639748/4c3dee10-b3cc-11e6-9c3b-a2a104f046ac.png)

**Comment by cinnamoner** [comment]
@NikoKrause I do use your SarkMenu, but I don't use the Mate option in it. To me, the Mate layout is futile.

**Comment by cinnamoner** [comment]
*People, does any of you want the Mate layout in StarkMenu? Please speak up now!*

**Comment by dominichayesferen** [comment]
@cinnamoner Yeah, though, it would be even cooler to see that concept of the Tiled Menu simulation to be in that menu

**Comment by cinnamoner** [comment]
@feren That wasn't the question. Ugh! That's something else. 
***
Let's try again.
Do any of you want the Mate layout be part of StarkMenu?

[Y] / [N]

Be revengefull an hit yes. Hahaha  :) or keep your cool.

**Comment by cinnamoner** [comment]
Just kidding. Whatever the community desires. I've already stated my position.

**Comment by dominichayesferen** [comment]
Just doing some stuff with KDE Neon to try and convert KDE Neon into... err... what should I even call it, feren OS Neon?? (XD)

**Comment by cinnamoner** [comment]
@feren I've complied to the request of the community, namely to not post my stuff where it doesn't belong. I'm kindly asking you to cross over to Eldorado aka my thread (where anything goes). Thanks.

**Comment by BigDaddyLinux** [comment]
Wait.....is that cinnmoner asking someone to not go off-topic....LOL
Just kidding.
@cinnamoner obviously I am on KDE right now but when I use cinnamon I would like to have the mate layout there. 

> More isn't necessarily better 

As a KDE user, the options are what make KDE great, so I believe the more options you give people (unless it affects stability) the better off and the more people you can reach.

@feren I would love Feren OS Neon

**Comment by dominichayesferen** [comment]
@BigDaddyonTech Sadly I could only get a partial installation of KDE Neon into feren OS (partly Linux Mint, partly Neon), maybe if Neon Desktop was easier to install on Ubuntu...

**Comment by cinnamoner** [comment]
@feren ^ No offence but this thread is about StarkMenu :) I'm kindly asking you to cross over to Eldorado.
@NikoKrause The Mate layout doesn't boder me as long as it doesn't affect the the primary layout or doesn't bring shortcomings of any kind.

**Comment by dominichayesferen** [comment]
@cinnamoner Well, he mentioned about KDE...

**Comment by cinnamoner** [comment]
@NikoKrause Finally I've got the opportunity to install Stark on 2.8.8. Wanna replace Configurable in Lindoz with Stark.
@feren Do you anticipate problems as to my endeavour ^?

**Comment by cinnamoner** [comment]
@feren At quantum level all things are linked so ^ I understand ^ :)

**Comment by NikoKrause** [comment]
@Odyseus
I want to switch to multiversion. I found the article about it.
I saw you use `cinnamon-version` tag in `metadata.json` file. Can you explain me, what it does and what it doesn't?

**Comment by cinnamoner** [comment]
> I want to switch to multiversion.

@NikoKrause Fine. Please make sure 2.8.8 gets included as well. Thank you. 👍

**Comment by NikoKrause** [comment]
@cinnamoner 
Don't worry about 2.8.8
Have you tried on 2.6?

**Comment by cinnamoner** [comment]
@NikoKrause Not, I haven't tried it on 2.6. Sorry. At least not yet.
@feren Which version of Linux Mint features 2.6? If you don't know by heart, I'll run a search. Thanks.

**Comment by Elbullazul** [comment]
@cinnamoner I think 17.2 or 17.3, but you need to upgrade, and maybe they replaced the 2.6 tarballs with newer ones, so I don't know. Wikipedia says 17.2 is THE build

**Comment by ghost** [comment]
Hi people, some of you known if there are any python script or something similar, with the ability of read the metacity xml and generate the close, maximized minimized  buttons? I have problems with this, because i want to create something like this: https://github.com/fatihmete/window-buttons-with-title but generate the buttons base on the metacity xml and not fixed it by themes.

**Comment by cinnamoner** [comment]
> Wikipedia says

@Elbullazul Although I think Wikipedia is a noble and useful endeavour, that's not what I meant by running  a search :)

**Comment by Odyseus** [comment]
Hello, everybody.

@NikoKrause.
The `cinnamon-version` key inside the **metadata.json** file is there only to inform the user that the xlet might not be compatible with the version of Cinnamon he/she is using. It has nothing to do with the `multiversion` option. In short, use the `cinnamon-version` key to declare the Cinnamon versions that your applet has been tested on.

For example. In my Custom Cinnamon Menu applet I added:
```json
"cinnamon-version": [
    "3.0",
    "3.2"
]
```

People trying to install the applet on Cinnamon version 2.8.x will receive a warning like this one:

![cinnamon-version](https://cloud.githubusercontent.com/assets/3822556/20814104/16474f70-b7f7-11e6-9d16-6d5f48a9b052.png)

I had to do this because I don't have a system with Cinnamon 2.8.x. I only test in that version of Cinnamon very briefly with a live system (the Linux Mint 17.3 Cinnamon ISO). When I tested my menu applet in there, I found out that some parts doesn't work very well, so I kind of directly removed compatibility for those versions of Cinnamon. The thing is that I can't make the fixes to the applet inside the live system because I found the default software totally useless (to name the more important to me, Nemo and xed are unusable). And installing and configuring the software I need every single time that I have to test an xlet is a total bother.

**Comment by NikoKrause** [comment]
@cinnamoner 
Cinnamon 2.6 is LM 17.2 https://www.linuxmint.com/rel_rafaela_cinnamon_whatsnew.php
You don't have to test it there. I assumed you have a LM17.2 machine running.
The oldest I have is 2.8.8

@Odyseus 
Thanks for the explanation.

@everybody
Who of you is native speaker in English?

**Comment by Elbullazul** [comment]
@lestcape how I would do it:

- get metacity name (gsettings?)
- find buttons in theme
- write quick style.css like in applet (guess the states of buttons according on their names)
- copy images to new folder in applet named after theme

my guess is it would be easier to write a shell script instead

**Comment by ghost** [comment]
@Elbullazul thanks, i think you speak of something else... I speak of this:
https://wiki.gnome.org/Attic/GnomeArt/Tutorials/MetacityThemes
See the images bellow, this properties are inside of the metacity xml:

![](https://wiki.gnome.org/Attic/GnomeArt/Tutorials/MetacityThemes?action=AttachFile&do=get&target=art-tutorial-mcity-im1.png)

but thanks anyway.

**Comment by BigDaddyLinux** [comment]
@NikoKrause 

> Who of you is native speaker in English?

:+1:

**Comment by NikoKrause** [comment]
@BigDaddyonTech 
Since you're a native speaker, would you like to correct my English for StarkMenu settings and Readme?
Not now. But soon.

**Comment by cinnamoner** [comment]
> Who is a native speaker in English 

@BigDaddyonTech I'm sure you'll do a great job.
@NikoKrause *Who's a native speaker of English who also speaks English correctly?*
*Speaking in*
*Speaker of*

**Comment by BigDaddyLinux** [comment]
I am glad to help out where I can, I might need to a little time but I will work on it.

**Comment by Elbullazul** [comment]
@lestcape you want to import data from the XML file instead? like, all the border, color & stuff?

**Comment by dominichayesferen** [comment]
@NikoKrause Native here too, UK English

**Comment by dominichayesferen** [comment]
@Elbullazul By the looks of it, he wants that, and it to also detect the paths of the image resources where applicable and use those images for the buttons...

**Comment by ghost** [comment]
@Elbullazul not i only want to generate the buttons. What happens it's that not all themes have an icons, for example Adwaita, Atlanta, Bright... There are of course, themes where more less you can use his icons, like Ambiance, AgingGorilla... Why I try to do it's read the xml and execute the necessary task to create the buttons... For example in the case of Adwaita, this mean paint the buttons from the instruction that the xml contain.

**Comment by Elbullazul** [comment]
@lestcape yeah, but you can't do that only with CSS. Maybe you'll have to generate a couple of images (close, max, min etc. icons for buttons)

**Comment by ghost** [comment]
@Elbullazul yes, this is why i ask for a python script or some thing similar to read the xml, i never speak about css (was you who speak about this). This is why i say you are talking of something different, I know i can apply some css rules later, but that's nothing problematic to me, understand the xml it's.

**Comment by NikoKrause** [comment]
> BigDaddyonTech: I am glad to help out where I can, I might need to a little time but I will work on it.
> feren: @NikoKrause Native here too, UK English

@BigDaddyonTech @feren 
The more eyes the better :)
It's not needed to correct it right now, because I'm still rearranging and rewrite some stuff.
The time will come, when I rename the applet. That's the signal for the Beta stage (i.e. feature freeze and only critical bug fixes) and for you to get involved.
![time](https://cdn.meme.am/instances/500x/62045247.jpg)

**Comment by BigDaddyLinux** [comment]
:+1:

**Comment by Elbullazul** [comment]
@lestcape you want some help to understand the xml syntax & rendering? if it's that, no problem. I can help

**Comment by ghost** [comment]
@Elbullazul yes, thanks. But not now, there are not a directly way to read an xml in cinnamon. The feature to do that was removed from firefox core (the base of gjs) and there are not any replace yet. Also, the metacity xml are not well-former (http://www.w3schools.com/xml/xml_validator.asp), so they will need in some cases an string replace to create a well former xml to be possible read it (if you want to know when a metacity xml it's well former, just drag and drop the xml to firefox if will be displayed without error, it's well-former, otherwise not). When i have some thing that be able to read most of the metacity xml, i will open a new github repository and i will invite you to help. Again thanks.

**Comment by dominichayesferen** [comment]
@NikoKrause I'm having issues with making an extension that simply maximises everything for feren OS's Tablet Mode... I'm basing it off of gTile, but now I'm getting 'missing ; before statement' when I try to use it, I've uploaded it, though, if you can also make it only do it if the window is not tiled nor maximised, then please do that too:
[maximizeeverything@feren.zip](https://github.com/NikoKrause/CinnXPStarkMenu/files/629632/maximizeeverything.feren.zip)

**Comment by NikoKrause** [comment]
> if you can also make it only do it if the window is not tiled nor maximised, then please do that too

I don't understand this sentence.

**Comment by dominichayesferen** [comment]
@NikoKrause If you could make it so that if the target window is not maximised or is not tiled, then maximise it...

**Comment by ghost** [comment]
@NikoKrause please see... There are one thing that i considered really wrong with this menu... The panel it's not dynamically, you can not drag and drop applications/places to the panel (where you have the icon). I'm talking about what i called accessible panel in configurable menu: https://github.com/lestcape/Configurable-Menu/wiki/Windows-7 (this is the box 2). Yo can see what configurable menu can do with this panel here:  https://www.youtube.com/watch?v=c5ljXTbwZUo

The settings for this it's absurd, long and a complete miss... remember, this is my opinion.

**Comment by NikoKrause** [comment]
@lestcape 
You are right. It's not possible to drag&drop to the sidebar.
I don't know how to do that. I think the buttons in the sidebar have to be stored somewhere.
I don't even know, where the favorites are stored.

You are also right, that the settings for this is long. Which Version have you tried?
I splitted the settings for this into icon, label and command. Before that it was all in one line.
I had the feeling, that splitting it, would make it easier to create own buttons on the sidebar.

**Comment by ghost** [comment]
I download the applet here, for the master brash...

- What i do it's use cinnamon settings to store the application id and place id. One variable for applications and another for places. I use a separator and generate an string representation of the list to be store in one variable. For places: https://github.com/lestcape/Configurable-Menu/blob/master/configurableMenu%40lestcape/settings-schema.json#L833 For applications: https://github.com/lestcape/Configurable-Menu/blob/master/configurableMenu%40lestcape/settings-schema.json#L837

- I think it's not important enable the drag and drop support for now. Just add a context menu for places with an option to add a place to the sidebar. Add an option to the context menu for application to add an app to the sidebar. Then add a context menu in the buttons of the sidebar to remove an element. I think in this way will better that what you have.

- A drag and drop option can be implemented later, as it's a little complex, because the menu have two views and you will need to swap between views when a drag and drop begin.

**Comment by ghost** [comment]
Anyway, what you need to support the drag and drop it's repeat the code of the favorites box to be used in the sidebar, one time per application and another box for places. This is not complex taking in account that you have the necessary code. The favorites box it's exactly what you want for applications. For places it's the same, but you will need to change the filter inside the corresponding functions, for example: https://github.com/NikoKrause/CinnVIIStarkMenu/blob/master/CinnVIIStarkMenu%40NikoKrause/2.8/applet.js#L1841

**Comment by Odyseus** [comment]
Hello, everybody.

@NikoKrause.
I would like to add a couple of things to what **lestcape** said.

I recently had to deal with drag and drop functionality. I thought that it would be out of my league (like **lestcape** said, it's very complex), but it turned out to be super easy.

Basically, you just have to define **handleDragOver** and **acceptDrop** in the element that you want to accept a dragged element.

This is a very basic example:

```javascript
this._allAppsCategoryButton.handleDragOver = function(source, actor, x, y, time) {
    return DND.DragMotionResult.MOVE_DROP;
};
this._allAppsCategoryButton.acceptDrop = function(source, actor, x, y, time) {
    let app = source.app;
    let id = app.get_id();
    let favorites = AppFavorites.getAppFavorites().getFavoriteMap();
    let srcIsFavorite = (id in favorites);

    Mainloop.idle_add(Lang.bind(this, function() {
        let appFavorites = AppFavorites.getAppFavorites();

        if (srcIsFavorite)
            AppFavorites.getAppFavorites().removeFavorite(id);
        else
            AppFavorites.getAppFavorites().addFavorite(id);

        return false;
    }));

    return true;
};
```

This code adds drag and drop capabilities to a category. It's a simplified version of the function used by the Favorites box. Basically, I striped all positioning code.

For you to implement it on your sidebar's menu, you may need some more complex calculations because you may need to position the dragged object into an specific location. In addition, you will need to make the places button draggable because they aren't by default (application buttons and favorites buttons are draggable). It seems that making an object draggable is very simple too. If you look at the **FavoritesButton** and **ApplicationButton** prototypes, the following two lines is what makes an object draggable:

```javascript
this._draggable = DND.makeDraggable(this.actor);
this.isDraggableApp = true;
```

I don't know if the second line is really needed. Someone more qualified could shed some light (@lestcape?).

>  I don't even know, where the favorites are stored.

The favorites are handled by the **appFavorites.js** module and they are stored into a gsetting (org/cinnamon > favorite-apps key). It's an array with the favorites applications IDs (the name of their .desktop files).

This will list the favorites in a terminal:
```
gsettings get org.cinnamon favorite-apps
```

Or simply search for the **favorite-apps** key on DConf editor.

**Comment by Odyseus** [comment]
@NikoKrause

**About cinnamon-json-makepot:** I will add some observations here about this so we can talk freely.

> Odyseus said:
> The po folder should be at the root of the xlet directory and contain the translations for all the xlets versions (with multiversion enabled).

This should be this way because the xlet managers and the `cinnamon-json-makepot` command look for the **po** folder containing the translations only inside the root of the xlet directory and nowhere else. And that's why the `cinnamon-json-makepot` command should always be run from the root of the xlet directory and from nowhere else. Furthermore, everything is *attached* to the xlet UUID, which also es extracted from the root of the xlet directory (the metadata.json file). Then that xlet UUID is used to create the *real translation file* (UUID.mo file inside the locale folder).

**Comment by ghost** [comment]
@Odyseus you are ok.
this.isDraggableApp = true; it's not needed, just was added to enable/disable draggable actions between menu and panel launcher. Of course, if you want that panel launcher accept a draggable buttons, you will need to use it. See:
https://github.com/linuxmint/Cinnamon/search?utf8=%E2%9C%93&q=isDraggableApp

Another important things. There are tow types of drag and drops, one that come from clutter objects and another that comes from gtk/kde (GDKWindows).
The clutter implementation it's here:
https://github.com/linuxmint/Cinnamon/blob/master/js/ui/dnd.js
The GDKWindows implementation it's here:
https://github.com/linuxmint/Cinnamon/blob/master/js/ui/xdndHandler.js

Please see: the GDKWindows implementation finally use the clutter implementation (to be something like a generic drag and drop) and see the GDKWindows implementation just not add information about what it's dragging, so you can not know anything about the GDKWindows drag and drop, you only can know that there are something that it's in a drag and drop process. This is a fault as in my opinion it could be possible know more information.

Note also that GDKWindows drag and drop won't occurs in the same level of the desktop that the normal (clutter) drag and drop, this is because this come from a windows and the windows are in an special actor inside the stage. This also need to be fixed!!

Well that the basic. As what we use it's the normal drag and drop our attentions need to be in this implementation. So please note some important signals that are emitted from the DND:
**drag-begin:** https://github.com/linuxmint/Cinnamon/blob/master/js/ui/dnd.js#L252
**drag-end:** https://github.com/linuxmint/Cinnamon/blob/master/js/ui/dnd.js#L509 
**drag-cancelled** https://github.com/linuxmint/Cinnamon/blob/master/js/ui/dnd.js#L555

You can use it to know the drag and drop stage of a button.

**Comment by NikoKrause** [comment]
@Odyseus
Have you noticed:
The right click context-menu of applets has an item "Remove %s", and %s means the name of the applet. The name of default applets is translated, but not the name of third party applets, even though there is a translation in the po file.

I can't figure out why it's not translated? Do you have and ideas?

**Comment by Odyseus** [comment]
Hello, @NikoKrause.

I just checked one of my own applets and it happens exactly the same.

What I think that it's happening is that Cinnamon is using the **Gettext.gettext** (without arguments) function for translating strings. While in our applets we are using **Gettext.dgettext(appletUUID, string)**. I think that this is another of those things that Cinnamon developers did not think very well while they were implementing it.

It would be very easy to fix this annoyance from our end, we just need to override the original context menu entry. Something like the following will make it work right.
```javascript
this.context_menu_item_remove = new PopupMenu.PopupIconMenuItem(_("Remove '%s'").format(_(this._metadata.name)),
    "edit-delete",
    St.IconType.SYMBOLIC);
this.context_menu_item_remove.connect('activate', Lang.bind(this, function() {
    AppletManager._removeAppletFromPanel(this._uuid, this.instance_id);
}));
```

I tried to override just the label of that menu item like this:

```javascript
this.context_menu_item_remove.label.set_text(_("Remove '%s'").format(_(this._metadata.name)));
```

This should have worked, but it seems that this context menu item is created very late on applet loading, so it's always **null** on applet initialization. I will see if I can find something more effective.

**Comment by ghost** [comment]
How translations works:

1- There are one main domain translation and several secondary domain translations (per extensions).
2- In cinnamon the translations of the stock extensions are inside the main domain.
3- In the internal classes (like popup menu) are not used the secondary domain translations, just the main domain is used.
4- There are one internal class where the secondary domain translations are used. This is the about dialog class.

So, if you want to add support for other internals classes you will want to see the about dialog implementation, and you will need to create a pull request with your changes in the cinnamon side... Another way could be hack the code and inject your changes.

**Comment by Odyseus** [comment]
**Edit:** Oops. **lestcape** beat me to it with other words. LOL

@NikoKrause

I forgot to specify that the **Gettext.gettext(string)** function is getting the strings from the Cinnamon's domain (Cinnamon translations) while the functions that we are using (**Gettext.dgettext(appletUUID, string)**) extract the strings from the applet's domain (that's why we are passing the applet UUID).

**Comment by ghost** [comment]
This is set per files, so need to be set inside cinnamon ui files like popupmenu file if you want to translate the string:
https://github.com/NikoKrause/CinnVIIStarkMenu/blob/f4966056062f1b555589a21296ce0e4d3fec6d95/CinnVIIStarkMenu%40NikoKrause/2.6/applet.js#L1967-L1972

**Comment by ghost** [comment]
The most easy solution:

Add this to your constructor: https://github.com/NikoKrause/CinnVIIStarkMenu/blob/master/CinnVIIStarkMenu%40NikoKrause/2.6/applet.js#L1979
```
if(this.context_menu_item_remove)
    this.context_menu_item_remove.label.set_text(_("Remove '%s'").format(_(this._meta.name));
```

This will override the text of the label and like it's in your applet.js file this will be translated with your applet domain.

Also i fully recommended change this function:
https://github.com/NikoKrause/CinnVIIStarkMenu/blob/master/CinnVIIStarkMenu%40NikoKrause/2.6/applet.js#L1970

for this one:
```
function _(str) {
   let resultConf = Gettext.dgettext("CinnVIIStarkMenu@NikoKrause", str);
   if(resultConf != str) {
      return resultConf;
   }
   return Gettext.gettext(str);
};
```
because this will include the translation for the cinnamon domain. so, any default translation present in the main domain could be used...

**Comment by ghost** [comment]
If you want to get more priority to the main domain translation, i.e try to use more uniform translation for cinnamon you could implemented the function in this way:

```
function _(str) {
   let resultConf = Gettext.gettext(str);
   if(resultConf != str) {
      return resultConf;
   }
   return Gettext.dgettext("CinnVIIStarkMenu@NikoKrause", str);
};
```

**Comment by Odyseus** [comment]
Hello, everybody.

@NikoKrause
Like **lestcape** said, I also recommend you to change the **_()** function. Otherwise, you will be forced to include default Cinnamon strings in your translations (you can't leave them blank). Using the function specified by **lestcape**, you can leave default strings blank in your .po files. I use the exact same function in my applets since I saw it in one of **lestcape**'s applets. :wink:

@lestcape
You are missing a closing bracket on this piece of code (see the one surrounded by asterisks):

```javascript
if(this.context_menu_item_remove)
    this.context_menu_item_remove.label.set_text(_("Remove '%s'").format(_(this._meta.name))**)**;
```

But that code doesn't work for me. At least not on my systems. Because, like I said in my previous message, **this.context_menu_item_remove** is always **null** in applet initialization. If I put that inside a **timeout**, then it works. Like this:

```javascript
Mainloop.timeout_add_seconds(5, Lang.bind(this, function() {
    if (this.context_menu_item_remove)
        this.context_menu_item_remove.label.set_text(_("Remove '%s'").format(_(this._meta.name)));
}));
```

I gave it 5 seconds just to be sure and because it isn't that important for that item to be available too early.

But, like you said in one of your previous messages, this should be fixes on Cinnamon's side. I will start working on a fix. If I'm successful, I will have a pull request ready by the end of the day.

**Comment by ghost** [comment]
**Edit:** _mmmm. I don't know why you need the *._ Yes i see what you say...
And yes sorry, finalized it's called by applet manager, so, the propretly way will be add this function to the applet implementation, thats will override the internal implementation:
```
finalizeContextMenu: function () {
    Applet.Applet.prototype.finalizeContextMenu.call(this);
    if(this.context_menu_item_remove)
        this.context_menu_item_remove.label.set_text(_("Remove '%s'").format(_(this._meta.name));
},
```

**Comment by Odyseus** [comment]
@lestcape
> mmmm. I don't known why you need the **.

I placed the asterisks just to point you where the bracket was missing (as a visual guide). Of course that they aren't needed.

**Comment by ghost** [comment]
@Odyseus not we are both wrong, when you move the applet to another panel, again will be called the finalizeContextMenu, so we need to ensure that always will set our translations... there are no way to call this directly in the constructor or call it in a time out... The only way will be override the finalizeContextMenu like my previews comment.

**Comment by Odyseus** [comment]
@lestcape
> when you move the applet to another panel, again will be called the finalizeContextMenu

That's easy to fix. We can just put the code inside the **on_orientation_changed** function and problem solved. The thing is, we might need to implement this *hack* because, even if I push the fix to Cinnamon, they may not implement it until the next version (3.4).

I already pushed the commit to my fork ([Give it a look](https://github.com/Odyseus/Cinnamon/commit/5a9f2a1309ab87bd04996ceb746dd99b5afd27e8) if you can). It works, but I don't know if it's the "accepted way to go".

**Comment by cinnamoner** [comment]
@raymerjacque The finest and most knowledgeable people in issues related to Cinnamon applets are right here.

**Comment by ghost** [comment]
@Odyseus, yes if you find all places where this function is called, you can fixed, but what about when another call for other place will be done... in the main code because a change or in an extension because one people want... You can not think to patch all... It's better do the thing in a good way...

For example your implementation for the translations are not good... This is because you not preserve the common signature for a translation, you can not execute the python script that you fixed, for extract the translation automatically, because this script use a c program and you can not be sure that **this._()** will be interpreted as a translation because have not the same signature **_("....")**.

**Edit:** What i say it's:  **_("....")** are not the same of **this._("....")**

**Comment by Odyseus** [comment]
@lestcape: It doesn't really matter if **this._()** isn't recognised by the *string extractors*. The **this._()** call isn't meant to be extracted, it's just meant to translate a string on the fly. And I used the exact same method used by the applet's about dialog that you yourself pointed out. [Here it's called](https://github.com/linuxmint/Cinnamon/blob/master/js/ui/modalDialog.js#L479) and [here it's defined](https://github.com/linuxmint/Cinnamon/blob/master/js/ui/modalDialog.js#L562).

> What i say it's: _("....") are not the same of this._("....")

Of course that they aren't the same. And like I said, it matters not.

**Edit:**
I was thinking of moving the **Gettext.bindtextdomain()** call into **_init**. So it's called only once on applet initialization and not every time **finalizeContextMenu** is called. What do you think?

**Edit 2:**
On second thought, it wouldn't be wise to move **Gettext.bindtextdomain()** into **_init**. I just saw that **this._uuid** is **null** inside **_init**. Who knows at what point that is set.

**Comment by ghost** [comment]
1- Gettext.bindtextdomain need to be called at the begin of your script and just one time...
2- It's not the same read the extension translation domain in a cinnamon code than read the extension domain in the extension code. The example of the about dialog it's if you want to change things inside the cinnamon classes, where you don't need to follow a general translation mechanism. Inside the extension the best it's do the things as it's usually.

**Comment by ghost** [comment]
@Odyseus I think this function name it's not necessary and it's confused: https://github.com/linuxmint/Cinnamon/pull/6118/files#diff-f166c2bedbb381ede2230178e39b8bc1R535 but well... I hope this will interpreted ok by the translation extractor of cinnamon... Other thinks are ok to me.

**Comment by Odyseus** [comment]
@lestcape: I don't like that function name either, but for different reasons. You should stop worrying about the "translation extractors". For one simple reason. **xgettext**, the one that does the actual string extraction from the sources, ignores any call to the **_()** function (or equivalent) that doesn't contain a *pure string*. For example, **_("string")** will be extracted by **xgettext**, but **_(expression)** will be completely ignored. For this reason, even if **xgettext** were too dumb to confuse **this._()** with **_()**, the function that I added and the one that I based it on uses it like this **this._(expression)**. So, errors are simply not possible.

**Comment by ghost** [comment]
Ok, if you say, but i don't like it... 
Several years ago mtwebstern also disagree to create the infrastructure to translate things for the applet domain translation inside cinnamon... He say not for now, maybe will be now the time to do it... i don't know.

**Comment by cinnamoner** [comment]
https://youtu.be/Ishr_4A3xkk

Configurable on Cinnamon 3.2.6.

**Comment by NikoKrause** [comment]
What is the difference between this
```js
function _(str) {
   let resultConf = Gettext.dgettext("CinnVIIStarkMenu@NikoKrause", str);
   if(resultConf != str) {
      return resultConf;
   }
   return Gettext.gettext(str);
};
```

this
```js
function _(str) {
   let resultConf = Gettext.gettext(str);
   if(resultConf != str) {
      return resultConf;
   }
   return Gettext.dgettext("CinnVIIStarkMenu@NikoKrause", str);
};
```

And the one I use right now?

**Comment by Odyseus** [comment]
Hello, @NikoKrause.

The first function looks for a translated string in your applet's domain first and then it looks on Cinnamon's domain as a "fallback". The second function will look in Cinnamon's domain first and then it looks on your applet's domain as a "fallback". As @lestcape explained in [this comment](https://github.com/NikoKrause/CinnVIIStarkMenu/issues/9#issuecomment-268260370), the second function will give priority to translated strings found on Cinnamon's domain, ignoring the translated strings from your applet's domain.

The function that you are using currently...

```javascript
function _(str) {
    return Gettext.dgettext("CinnVIIStarkMenu@NikoKrause", str)
}
```

...doesn't have a "fallback". In case there isn't a translated string, the original string (**str** argument) will be returned.

**Comment by Odyseus** [comment]
@NikoKrause.

Just to clarify, the function that you are using isn't *bad* or *erroneous*. The thing is that, for example, you can't do what I do in the translation files of my applets. In my case, I could simply ignore (leave blank) the translation of certain strings that are already found on Cinnamon's domain, because I use the **`return Gettext.gettext(aStr);`** "fallback".

Another thing. When you declare the **_()** function, you are **overwriting** the default/existent **_()** function. The original **_()** function is just an abbreviation for **Gettext.gettext**. By overwriting the original function with the one that you are using, you loose the capability to fetch the translated strings found on Cinnamon's domain. By using either of the functions that you posted in your example, we are **expanding** the original function's capabilities.

**Comment by ghost** [comment]
@NikoKrause, @Odyseus  explain perfectly the difference, but i will add some examples:

if you use the seconds function, and you have a translation in your domain, and cinnamon have a translation in his domain, i think will be preferable use the cinnamon translation, because the translation it' s used in other places of cinnamon, so will be more uniform for users... Also if you have not a translation for some language, but you use the same strings as cinnamon and cinnamon have this translation, this string will be show in the languages of the user, and this is better that do not have translation at all. At lees it's what i think...

**Comment by NikoKrause** [comment]
Thanks for the explanation. I try it.

**Comment by NikoKrause** [comment]
I tried to understand your _() function and created an example.

Cinnamon and my applet are using the same string: `"Show application icons"`
In the German po file it looks like this
```
msgid "Show application icons"
msgstr "Anwendungssymbole anzeigen"
```

I changed it in my applets german po file to this: 
```
msgid "Show application icons"
msgstr "Anwendungssymbole"
```

No matter which _() function I use. Both time the String from the applets po file is used.

**Comment by NikoKrause** [comment]
Also if I leave the translation blank, it's still translated from the Cinnamon translation, even with my old function.

**Comment by ghost** [comment]
@NikoKrause you are testing things inside the settings? This function only have an scope that is limited to the current js file.

**Comment by Odyseus** [comment]
@NikoKrause.

I'm going to expand what **lestcape** said. The *"Show application icons"* string is inside your *settings-schema.json* file. The **_()** function that you declare inside your .js file doesn't translate strings located inside that settings file. The **_()** function is only called inside your .js file, so it will translate strings found only inside your .js file.

The strings found inside the *settings-schema.json* files are translated on the fly by [this Python function](https://github.com/linuxmint/Cinnamon/blob/master/files/usr/share/cinnamon/cinnamon-settings/xlet-settings.py#L51) every time the settings windows are opened. As you can see on that function, translated strings defined by your applet will have priority over the strings translated by Cinnamon.

**Comment by NikoKrause** [comment]
@Odyseus 
> Spices comments system is absolutely useless to report bugs with any king of legibility. In addition, there is no notifications system for new comments.

Check this out:
https://github.com/collinss/Spices-Notifier

**Comment by Odyseus** [comment]
Hello, @NikoKrause.

Yes, I used that applet for some time and I also made some tweaks to add to it some features. But I abandoned it after hitting a wall with the **Soup** API (it's an API to make HTML requests from JavaScript code). Basically, I found out that the **Soup** API is very limited for what I would like to do.

I'm waiting for the new Spices web site to come on-line and see if it is still needed an applet like the one you linked to handle notifications. I already gained some experience with a Python module to handle advanced HTML requests and I have in mind what I would like to do with it.

**Comment by NikoKrause** [comment]
I figured out, what the translation function from lestcape does.
It's indeed better.

@lestcape @Odyseus 
I'm using ngettext to show number of search results in my menu.
It's not translated. Any ideas on that?
https://github.com/NikoKrause/CinnVIIStarkMenu/blob/master/CinnVIIStarkMenu%40NikoKrause/3.2/applet.js#L4139

**Comment by NikoKrause** [comment]
Nevermind. I have to use dngettext instead of ngettext.

**Comment by ghost** [comment]
@NikoKrause why you use dngettext or ngettext directly? This not help when you want to extract all translation automatically.

**Comment by NikoKrause** [comment]
Yes. The strings are extracted as well.
At least with ngettext it was extracted. I haven't tested with dngettext, yet.

**Comment by ghost** [comment]
Well, i never tested it... My error... But in any case i think it's more readable if you have all in the same format... Of course it's like you want or how this look like better for you.

**Comment by NikoKrause** [comment]
How else can I use ngettext/dngettext?

**Comment by ghost** [comment]
You already have a function for translate things: https://github.com/NikoKrause/CinnVIIStarkMenu/blob/master/CinnVIIStarkMenu%40NikoKrause/3.2/applet.js#L1941 that internally use what you want. If you want to use a different mechanism or different Gettext internal function, just change this function like you want, then just simple use the function in your code like any translation like this one:
https://github.com/NikoKrause/CinnVIIStarkMenu/blob/master/CinnVIIStarkMenu%40NikoKrause/3.2/applet.js#L4137

In fact d**n**gettext, it's used when you want a different **main** translation domain, but if you use the default domain, why you need to change the function...

**Comment by NikoKrause** [comment]
Ah. I see, you have to know, ngettext is used for something different.

Many languages have just Singular and Plural.

For example.
If you have 1 tomato, you say `tomato`.
If you have more than one tomato, you say `tomatoes`.

Some languages have more than one plural.
They use depending on the number different Plurals.

1 tomato
2, 3, 4 tomati
5, 6, 7, 8, 9 tomatitos

So you can't just use the normal gettext/dgettext.
And that's what ngettext is for. It handles that for you.

**Comment by ghost** [comment]
@NikoKrause thanks for that, this is new for me... and why then you don't change this function: https://github.com/NikoKrause/CinnVIIStarkMenu/blob/master/CinnVIIStarkMenu%40NikoKrause/3.2/applet.js#L1941

For this one:
```
function _(str) {
    let customTranslation = Gettext.dngettext(UUID, str);
    if(customTranslation != str) {
        return customTranslation;
    }
    return Gettext.dngettext(str);
}
```

**Comment by ghost** [comment]
This have side effects on other places or something?
I think it's preferable if have not side effects in other places, use a general mechanism for translations instead of specific cases...

**Comment by NikoKrause** [comment]
ngettext haa a different Syntax.

gettext("tomato")
ngettext("%d tomato", "%d tomatoes", number)

**Comment by ghost** [comment]
mmm, I see... thats good. I never use a thing like that.... Also, this is the first time that i see this... thanks @NikoKrause

**Comment by NikoKrause** [comment]
Yeah. I didn't consider it at first.
Many languages a French, English, German, Spanisch, etc. have one singular and one plural.
https://translations.launchpad.net/linuxmint/latest/+pots/cinnamon/es/+details
But languages as Ukrainian, Russian, etc. have one singular and two plural forms for such things.
https://translations.launchpad.net/linuxmint/latest/+pots/cinnamon/uk/+details

**Comment by NikoKrause** [comment]
@mate-desktop @raveit65
Can you please tell me, which License this icon of yours has?
https://github.com/mate-desktop/mate-desktop/blob/f543545ceb97db7cdefc186889898e3ebdef8f65/icons/hicolor_apps_scalable_mate.svg

I mixed it with the Cinnamon Spices icon http://segfault.linuxmint.com/wp-content/uploads/2016/09/CinnamonIcon5-1.png
And I want to use it for my Cinnamon Menu Applet. So I need to know, which Licence it is.

**Comment by Odyseus** [comment]
Hello, g@ys.

@lestcape & @NikoKrause: There is no need to modify the **_()** function that we are using to support the plural forms. If you use something like the following...
```javascript
let myPluralForm = dngettext(aMetadata.uuid, "%d result found", "%d results found", numberResults).format(numberResults);
```
...it will be normally extracted by the **cinnamon-json-makepot** command without the need to specify a keyword.

Or you could simply create another function called **dngettext()**.
```javascript
function dngettext(aSingular, aPlural, aN) {
    let customTrans = Gettext.dngettext(appletUUID, aSingular, aPlural, String(aN));

    if (customTrans !== aSingular && customTrans !== aPlural)
        return customTrans;

    return Gettext.ngettext(aSingular, aPlural, aN);
}
```

What I observed is that the function call has to be named **dngettext** or **ngettext**, otherwise, the string extractor will not recognize its usage as a plural form for **Gettext**.

For example, if the function above is called **N_** (Cinnamon uses this name in a couple of places) instead of **dngettext** and is used like this...
```javascript
let myPluralForm = N_(aMetadata.uuid, "%d result found", "%d results found", numberResults).format(numberResults);
```
...the string extractor will not recognize it at all, even if I pass the correct argument to scan for the **N_** keyword.

Another thing that I noticed is that I have to pass the number (third argument) as a string. Otherwise, I got an error complaining that the number is a number (¬¬).

**Comment by Odyseus** [comment]
Hello, everybody.

@lestcape, @NikoKrause and anyone interested.

It seems that tomorrow the new Spices website will see the light. Clem has been uploading what seems to be the repositories which we (xlet developers) will have to use to push pull requests to update our xlets. The followings are the new repositores.

- [**cinnamon-spices-applets**](https://github.com/linuxmint/cinnamon-spices-applets)
- [**cinnamon-spices-extensions**](https://github.com/linuxmint/cinnamon-spices-extensions)
- [**cinnamon-spices-desklets**](https://github.com/linuxmint/cinnamon-spices-desklets)
- [**cinnamon-spices-themes**](https://github.com/linuxmint/cinnamon-spices-themes)

I just hope for they to add somewhere some instructions to make *sparse checkouts*. Because cloning the entire freaking repositories to update just one xlet will be absolutely annoying. The themes repository alone is more than 400MB!!! I'm already prepared for this, so I have already investigated [how to clone just a part of a repository](http://stackoverflow.com/a/28039894). Although, I didn't tried it yet.

**Comment by ghost** [comment]
@Odyseus if i don't understand wrong the Clement words, the idea it's that: This will be update from our own repositories automatically. When we push an update in our master branch, the update will be automatically pushed to the cinnamon-spice-* repository. If i will need to update the cinnamon spices repository, like you say, well user will need to wait more than his live to be occurs for my applets....

**Comment by Odyseus** [comment]
@lestcape: What it's updated automatically is the Spices website from the *cinnamon-spices-** repositories. We have to fork and clone the *cinnamon-spices-** repositories and create pull requests to update our xlets.

> When we push an update in our master branch, the update will be automatically pushed to the cinnamon-spice-* repository.

That is simply not possible and will never be. Because, for example, my repository contains all my xlets. And even if it were possible, I wouldn't want for that to happen, because what I upload to my repositories are beta versions. Only after some testing I publish them in the Spices website.

But none of what we are discussing here is really relevant. We have to wait for all the system to be up and running to see how much we are going to hate it or love it. LOL

**Comment by ghost** [comment]
@Odyseus of course it's possible, if you follow a package format in your repository. I only disagree about the fact that cinnamon will gain the development credit for the three-party work. I also don't want to be **forced** to contribute with my code or what will be that? I contribute with the thing that i want and where i want. 

- I have not any problem to donate the final product.
- I have not any problem to donate my time.
- I have not any problem to donate the code.

But i have a big problem if all code that i create it's now for other peoples and the credits it's for other people, and the problems it's for me. I don't know if you understand my point. But anyway if you have right, this will be the cause that i will removed all my things from spices.

I'm not a developer of cinnamon and i don't want to be a cinnamon developer.

**Comment by Odyseus** [comment]
@lestcape

> of course it's possible, if you follow a package format in your repository.

And that's part of why I said why isn't possible. I'm not willing to change the format of my repositories. I create my repositories to fit my needs and work flow, not the other way around.

Plus, the main reason for this move to GitHub is that the code that we (xlet developers) push to our xlets can be reviewed by others in search of errors and malicious code. So, the part were you comment...

> When we push an update in our master branch, the update will be automatically pushed to the cinnamon-spice-* repository.

...will simply not happen. And like I said before, I wouldn't want to happen.

My *worries* have nothing to do with credits (I don't really care about that). What worries me is that someone might want to make one of my applets compatible with a version of Cinnamon that I simply don't have nor want (so it will be impossible for me to test it or maintain it). Or someone might want to modify or remove a feature from one of my xlets just because it's *too dangerous* for a newbie. That is a headache that I simply don't want to deal with. So, if any of that happens, I will be forced to remove everything from Spices and change the licencing of all my xlets.

**Comment by ghost** [comment]
@Odyseus of course you have different priorities. For my priorities I also don't need to wait, I'm pretty sure that i don't want my code in a Mint/Cinnamon github repository. I can accepted this, just if this will be an automatically process.  I try to remove my xlet from spice, i can not do that right now. Will need to be later.

**Comment by ghost** [comment]
And @Odyseus if your inconvenient will occurs,I will explicitly deny the use of my name in any attempt to redistribute the software, when some one different change the code.

**Comment by NikoKrause** [comment]
The new spices website will pull the xlets automatically from
github.com/linuxmint/cinnamon-spices-*

This shouldn't change thinks much for the developer. You will get the credit for your applet. I think you can see, who uploaded the xlets on the spices site, like it always was.

The thing, that changes is, that instead of uploading a .zip file on spices, you have to create a pull request on github.
The code won't be reviewed line by line. That seems very time consuming and wouldn't allow much time for development of cinnamon, etc.
It'll be looked over, if it contains malicious code.

If other people than the owner/creator of the applet would want to change something on the applet, e.g. removing/adding New features etc.,
they have to create a Pull Request. And the pull request needs the approval of the owner/creator before it gets merged.

It should be possible to continue to work in your own repo. If you are happy with your work, you can copy/upload the files in the cinnamon-spices-* repo and create a PR.
Not very different than creating a ZIP file and upload that.

**Comment by NikoKrause** [comment]
@lestcape
> i don't want my code in a Mint/Cinnamon github repository

What's the difference in having your code on a mit github repo than in having your code on a mit Website, which is Cinnamin Spices?

**Comment by NikoKrause** [comment]
@Odyseus 
https://github.com/linuxmint/Cinnamon/issues/6130#issuecomment-274027436

**Comment by NikoKrause** [comment]
@Odyseus @lestcape 
https://cinnamon-spices.linuxmint.com/

The new spices site is online. It looks very good, except the themes, yet.
I hope it's not as bad as you fear. There is coming a notification on http://segfault.linuxmint.com/, which will explain, how things will work.

It's still your right to delete your applets, if you want to. But I think it won't be as that bad as you think.
We will soon see how the whole process, evolves.

@Odyseus 
> So,` if any of that happens, I will be forced to remove everything from Spices and change the licencing of all my xlets.

You can delete all your xlets, if you want to.
But I see problems in changing the licensing of ALL your xlets.
Some of them are forks from other applets under GPL, e.g. the Custom Cinnamon Menu.
So it has to be GPL too.

If you don't have written your xlet from scratch, you just can't use any another or even "closed" license.

**Comment by ghost** [comment]
@NikoKrause spice was a distribution site, owner by the autor like it's launchpad. Github it's a development site owner by the autor. The autor in spice was me. The author in github was me.

The credit of development a thing in github was for me and the credit for packing an applet in spice was for me and for linuxmint.

The credit of development in the repository of other organization will be for this organization, not for me. As the repository it's for LinuxMint organization the credit will be for mint. So, i will lost my development credits in favour of LinuxMint. I don't use LinuxMint and i dislike LinuxMint and i don't want to be a part of LinuxMint.

Be sure a repository of Mint will be more visible that my own repository. So, my work will be as a fork of Mint, when it's not in that way.

**Comment by ghost** [comment]
For those how want to see where I'm forking cinnamon at the top of gnome shell: https://github.com/lestcape/Classic-Gnome

**Comment by Odyseus** [comment]
@NikoKrause.

In your last 4 messages on Cinnamon issues you are referencing a **joseph** when I'm 99% sure that your intent is to reference **JosephMcc**. Am I right?

**Comment by NikoKrause** [comment]
Upps. You are right. Thanks.

**Comment by NikoKrause** [comment]
@Odyseus
I wondered if it's really needed, that you have localizations.sh files in each of your applets.

Wouldn't it suffice to say, use cinnamon-json-makepot -i in the Readme?

**Comment by Odyseus** [comment]
@NikoKrause

I prefer it that way. One can install/remove a localization from that script and also restart Cinnamon. Furthermore, it forces to launch the commands inside the applet's folder. A thing that some people seem to still not freaking understand! It seems that is not clear enough to specify "run the freaking command from inside the applet's folder". LOL

**Comment by NikoKrause** [comment]
Maybe those people shouldn't install it manually ;)

**Comment by Odyseus** [comment]
Hello, @NikoKrause.

Can I suggest you to move [the tutorial that you just created](https://github.com/NikoKrause/CinnVIIStarkMenu/issues/19) into your Wiki section? You will work much more comfortable and can be expanded easily. Where it is right now, a closed issue, is "hard to find" and kind of misplaced. In case that you don't know, [it's really easy to create a Wiki](https://help.github.com/articles/about-github-wikis/).

**Comment by NikoKrause** [comment]
You are right. I was too lazy to start a wiki.
I think it would make more sense to write this kind of tutorial here:
http://developer.linuxmint.com/reference/3.0/cinnamon-tutorials/extension-system.html

**Comment by Odyseus** [comment]
I think that it would make even more sense to write it in the [Cinnamon's repository Wiki](https://github.com/linuxmint/Cinnamon/wiki). The place that you linked is absolutely annoying to work with, because its source is [XML code](https://github.com/linuxmint/Cinnamon/blob/5bbf2f72d6fe4fe8611976a0e802b7a4ee26bec3/docs/reference/cinnamon-tutorials/xlet-versioning.xml).

**Comment by ghost** [comment]
@Odyseus Like i promised, I will "migrate", my applets and desklets to the Gnome shell desktop soon... They will not change the structure, but probably will not work in Cinnamon because i will uninstall it, so i will don' t know if they will be running or not in cinnamon, and also i don' t care to know, i will not fix it. If you want fork any other thing thats the time.

**Comment by Elbullazul** [comment]
@lestcape your unity menu to cinnamon will get kind of useless...

**Comment by ghost** [comment]
@Elbullazul what you say???? What unity Menu?? Why will get kind of useless...?

You need to see how it's working the fork of cinnamon in gnome shell first... It's more faster than cinnamon, have Wayland support, and support for the gnome extension also... Anyway, i will use my things and thats it's the most important things.

**Comment by ghost** [comment]
Ahhhh  @Elbullazul you say the global menu, ja... Yes i also don't like it, but well see: When you will have 48 start and 10 fork of your themes in github, like this stupid menu, we can speak again about of useless....

**Comment by Odyseus** [comment]
Thanks for the heads up and your consideration, @lestcape. I already forked all the repositories that I could find useful in the future. I did it mostly for reference. I look at others people code to learn how things work, and I learned a lot from you. Thank you very much for that too.

**Comment by Elbullazul** [comment]
@lestcape no, I wasn't meaning it IS useless, I was just saying it's useless to port it to gnome since it's a native feature...

my bad, my english is crappy by moments...

**Comment by germanfr** [comment]
@NikoKrause regarding your small tutorial. Why is it needed to define the `_` function? Isn't it supposed to be already defined by the system?

In [this applet](https://github.com/linuxmint/cinnamon-spices-applets/blob/ba0627d3dc21f86994771dc65a5022889741a662/system-monitor%40pixunil/files/system-monitor%40pixunil/applet.js#L16), for example, it's comming from somewhere else:
```js
const _ = imports.applet._;
```
And in the default preinstalled applets it's not even imported/defined anywhere.

It may be a silly question, sorry for the ignorance.

**Comment by Odyseus** [comment]
@germanfr: The default **_()** function extracts translated strings from Cinnamon's domain. The custom function specified by NikoKrause's tutorial extracts the translated strings from the applet's domain. Although, I wouldn't use that function because it overrides the original and there is no "fall back". I would recommend using the following instead:

```javascript
function _(aStr) {
    let customTrans = Gettext.dgettext(appletUUID, aStr);

    if (customTrans !== aStr)
        return customTrans;

    return Gettext.gettext(aStr);
}
```

**Comment by ghost** [comment]
@Elbullazul sorry then, what i say not was nice then... I really don't know if some will want that, i have 2 request to be migrate it to gnome shell and also you can find in the web a lot of people asking for it. What gnome have it's not a global menu it's like a compact menu.

**Comment by ghost** [comment]
This is what i currently have, if some one it's interested:
At left the cinnamon settings(python + Gtk), at right a settings migration to gjs + Gtk. 
At top the gnome components, at bottom, the cinnamon components.

 ![captura de pantalla de 2017-02-02 18-39-51](https://cloud.githubusercontent.com/assets/1389572/22574819/0b43fd10-e977-11e6-801d-10d5e3460993.png)

**Comment by germanfr** [comment]
Thanks @Odyseus. So if I understood, the difference is where the translated strings are taken from. I'll use that then (if you don't mind).

**Comment by Odyseus** [comment]
@germanfr.

> So if I understood, the difference is where the translated strings are taken from.

That's right.

> I'll use that then (if you don't mind).

You don't even have to ask. BTW, I learned that from @lestcape. 👍

**Comment by ghost** [comment]
@Odyseus in the last week i auto-forzed me to learn a lot more.... 

do that:
window._ = function(str) {
 ...your translation mechanims....
}

Then all cinnamon will use your translation mechanism in all places when you do:
_("") this is necesary to merge the cinnamon translation with the gnome shell translation in a fancy and transparent way.

**Comment by NikoKrause** [comment]
Hey, open this website (http://developer.linuxmint.com/index.html) and type `conway` :)

**Comment by NikoKrause** [comment]
Time to move this applet and this Chatroom to No. 3....
